Expose some new configuration values (CRON tasks, socket timeouts) (main)

alanking opened the pull request:
Addresses #6342 
Addresses #6370 

Core tests passed on ubuntu:20.04.

Couple of notes:
1. Open to changing any and all names and default values for the configurations as well as their affiliated variables in the code.
2. I thought about making `get_advanced_setting` a template function based on the type of the configuration value, but this seemed a little overboard considering everything is just `int`s (for now) and the function is only available in the one file so it doesn't need to be so general. Any thoughts on this would be appreciated.
3. I'm happy to hold on merging this to yield to #6673 due to impending conflicts.
4. Ignoring the clang-tidy issue because it's just getting confused by the generated header file for the control plane commands.

alanking commented:
Oh, also, I did *not* add any automated tests for this mainly because they would be timing based by definition and kind of hard to detect anyway. I can certainly do this if it seems helpful.

I confirmed at the bench that the configurations have the intended effects (except for the socket timeouts).

korydraughn commented:
I haven't looked over this yet, but some immediate things come to mind that we need to do:
- Verify upgrades from 4.3.0 to 4.3.1 add the new properties to server_config.json if missing
- Update documentation for server_config.json in irods/irods_docs repo
- Explain the ZMQ timeout options for send/recv

alanking commented:
Right, upgrades. I'll take a look at that.

ZMQ timeout options will be documented along with the other options in irods_docs (see checkboxes in linked issues). Is that what you meant by "explain" or is there something more?

korydraughn commented:
Yes. That is what I meant.

alanking commented on packaging/server_config.json.template:
```diff
@@ -10,13 +11,18 @@
         "maximum_temporary_password_lifetime_in_seconds": 1000,
+        "migrate_delay_server_sleep_time_in_seconds": 5,
+        "migrate_delay_server_socket_receive_timeout_in_milliseconds": 5000,
+        "migrate_delay_server_socket_send_timeout_in_milliseconds": 500,
```
This made me wonder if we should include a "delay_server" section in the configuration since we now have the following configurations:
 - delay_rule_executors
 - delay_server_sleep_time_in_seconds
 - maximum_size_of_delay_queue_in_bytes
 - migrate_delay_server_sleep_time_in_seconds
 - migrate_delay_server_socket_receive_timeout_in_milliseconds
 - migrate_delay_server_socket_send_timeout_in_milliseconds
 - number_of_concurrent_delay_rule_executors

alanking replied:
I suppose some of these are configurations for the delay server itself and some are configurations for machinery around the delay server, i.e. delay server migration... Anyways, something to consider.

korydraughn replied:
That's a possibility. The main challenge is upgrades and expectations of administrators.

There's also the discussion of a flat configuration vs a configuration with nesting.

alanking replied:
Very true. Keeping the flat structure for these options in particular seems fine/preferred, so I'll leave it how it is.

trel commented on packaging/server_config.json.template:
```diff
@@ -10,13 +11,18 @@
             "shared_memory_size_in_bytes": 5000000,
             "eviction_age_in_seconds": 3600
         },
+        "dns_cache_clearer_sleep_time_in_seconds": 600,
```
i like a flat hierarchy for these settings i think.   less complicated.  less logic to parse.

korydraughn replied:
Could that lead to more confusion given we have a section dedicated to dns_cache?

trel replied:
yes, probably.   bah.

alanking replied:
I think grouping this into the `dns_cache` and `hostname_cache` sections makes the most sense because the JSON objects already exist and it reduces the number of keywords we need to add. Plus, as we discussed, it would logically group the concepts together so that it is known that they are related.

korydraughn replied:
Go for it.

alanking replied:
Done

alanking commented:
The configurations are moving around a little bit, so I'm going to put this in draft. Feel free to review anyway

alanking commented:
Regarding upgrades...

Is my understanding correct that injecting the new values would require a schema version bump? 

Every new setting that I've added is considered optional, and some of the settings are embedded in structures that are themselves optional (`dns_cache` and `hostname_cache`). So we may need to discuss the rules around this should we desire the values to be injected.

trel commented:
new things don't require a version bump - as optional/new items will still be valid according to our current schema ( we aren't leveraging additionalProperties... https://json-schema.org/understanding-json-schema/reference/object.html#additionalproperties )

alanking commented:
Sorry, by "injecting" new values, I meant... adding the new values to an existing server_config.json on upgrade.

So it sounds like our upgrade logic provides a mechanism for doing this without bumping the schema version?

trel commented:
Yes, I believe we have done this since 4.2.0 on the 4.2.x series - and this should be the same, except perhaps the mechanisms have moved a little.  Defer to @korydraughn.

korydraughn commented:
This is the latest bit of code that was written to add missing properties. The `update_base` function could be pulled out and used when an upgrade is detected.
https://github.com/irods/irods/blob/2f602874739f3d0d15551127cddbac181fb88e27/scripts/irods/upgrade_configuration.py#L186-L199

I think adding properties without a schema version bump requires a little thought. Basically we need a mechanism similar to how schema version upgrades are handled. Or maybe the version last step in the process is a call to add missing properties. That may work.

alanking commented:
We have decided that the new configuration values will not be added on server upgrade because they are not required. See https://github.com/irods/irods/issues/7143.

Squashed and rebased. I think this is ready for review now.

Need to run tests again.

trel commented:
looks good upon initial inspection.

korydraughn commented on server/main_server/src/rodsServer.cpp:
```diff
@@ -606,8 +631,20 @@ namespace
+            const int timeout = [] {
+                try {
+                    return irods::server_properties::instance()
+                        .map()
```
Same here.

alanking replied:
Done

alanking commented on server/main_server/src/rodsServer.cpp:
```diff
@@ -376,26 +376,51 @@ namespace
 
-    int get_stacktrace_file_processor_sleep_time()
+    auto get_advanced_setting(const char* _keyword, int _default, bool _negative_values_allowed = false)
+        -> decltype(_default)
```
I think at the time I was playing around with the types and didn't feel like changing it in 3 places. I'll just make it an int.

alanking replied:
Done

alanking commented on server/main_server/src/rodsServer.cpp:
```diff
@@ -376,26 +376,51 @@ namespace
+    auto get_cache_clearer_sleep_time(const char* _keyword, int _default) -> decltype(_default)
+    {
+        try {
+            const auto adv_settings = irods::server_properties::instance().map().get_json().at(irods::KW_CFG_ADVANCED_SETTINGS);
```
Makes sense. Will do

alanking replied:
Done

alanking commented:
Force-push due to rebase. New changes are in separate commits.

Clang-tidy has two complaints which I'm willing to address, but neither are new to my changes. Let me know if these are important to you all:
1. "Cognitive complexity" of `main`
2. Magic numbers for the config values

---

Unit tests passed. Running core tests now.

korydraughn commented:
Neither of those clang-tidy checks matter much in regard to your changes IMO.

Will reiew the latest commits.

korydraughn commented on server/main_server/src/rodsServer.cpp:
```diff
@@ -631,8 +631,20 @@ namespace
             zmq::socket_t zmq_socket{zmq_ctx, zmq::socket_type::req};
-            zmq_socket.set(zmq::sockopt::sndtimeo, 500);
-            zmq_socket.set(zmq::sockopt::rcvtimeo, 5000);
+            const int timeout = [] {
```
Consider changing the variable name so that it's clear we're dealing with milliseconds.

Perhaps `timeout_ms`?

Or maybe we introduce a new config keyword as an alias. This new keyword would include `_IN_MILLISECONDS` or something to help readers understand the `10000`.

alanking replied:
Something like this?
```diff
--- a/server/main_server/src/rodsServer.cpp
+++ b/server/main_server/src/rodsServer.cpp
@@ -631,17 +631,18 @@ namespace
 
             zmq::context_t zmq_ctx;
             zmq::socket_t zmq_socket{zmq_ctx, zmq::socket_type::req};
-            const int timeout = [] {
+            const int server_control_plane_timeout_in_milliseconds = [] {
                 try {
                     const auto config_handle = irods::server_properties::instance().map();
                     return config_handle.get_json().at(irods::KW_CFG_SERVER_CONTROL_PLANE_TIMEOUT).get<int>();
                 }
                 catch (...) {
-                    return 10000;
+                    constexpr int default_server_control_plane_timeout_in_milliseconds = 10000;
+                    return default_server_control_plane_timeout_in_milliseconds;
                 }
             }();
-            zmq_socket.set(zmq::sockopt::sndtimeo, timeout);
-            zmq_socket.set(zmq::sockopt::rcvtimeo, timeout);
+            zmq_socket.set(zmq::sockopt::sndtimeo, server_control_plane_timeout_in_milliseconds);
+            zmq_socket.set(zmq::sockopt::rcvtimeo, server_control_plane_timeout_in_milliseconds);
             zmq_socket.set(zmq::sockopt::linger, 0);
             zmq_socket.connect(zmq_server_target);
             log_server::trace("Connection established.");
```

korydraughn replied:
Very clear. No one has to guess what is happening.

I think we have a file that defines global const[expr] default values. Perhaps the constexpr default can go there. Then it's available for other locations.

Sadly, I can't remember which file that is. Feel free to look for it, but if not, don't worry about it. What you've presented is certainly good enough.

alanking replied:
Hmm... if you can give an example for a default value that you have in mind, I may be able to locate it. So far, all default values I've found seem to be like this one - that is, local to where they are being used - or scattered throughout various header files which may or may not be related to the respective values. :)

korydraughn replied:
The thing I had in mind turned out to be specific to the delay server.

The delay server's hpp file contains constexpr variables. The function where I remember one of those being used was in a MSI that was recently removed.

All of that to say, I think you can move forward with the diff you presented.

alanking replied:
Okay. I say we leave it as this diff, then. Resolving

korydraughn commented on server/main_server/src/rodsServer.cpp:
```diff
@@ -398,10 +397,11 @@ namespace
         try {
-            const auto adv_settings = irods::server_properties::instance().map().get_json().at(irods::KW_CFG_ADVANCED_SETTINGS);
+            const auto config_handle = irods::server_properties::instance().map();
+            const auto adv_settings = config_handle.get_json().at(irods::KW_CFG_ADVANCED_SETTINGS);
```
`const auto&`?

I just realized that L384/385 also needs to be a reference.

alanking replied:
Yep, done

korydraughn reviewed (commented):
Just a couple things and then I think we'll be in a good spot.

alanking commented:
Unit and core tests passed

korydraughn reviewed (approved):
Please squash, pound, and merge when ready.
